{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "%run setup.ipynb\n",
    "%run peak-utils.ipynb"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "@functools.lru_cache(maxsize=None)\n",
    "def peak_scan(pop, chromosome, \n",
    "              filter_size=20, \n",
    "              filter_t=2, \n",
    "              gflanks=(4, 8), \n",
    "              scan_interval=1,\n",
    "              min_aic=500,\n",
    "              min_baseline=0,\n",
    "              max_baseline_percentile=95,\n",
    "              min_amplitude=0.03,\n",
    "              init_amplitude=0.5,\n",
    "              max_amplitude=1.5,\n",
    "              min_decay=0.1,\n",
    "              init_decay=0.5,\n",
    "              max_abs_skew=0.5,\n",
    "              diagnostics=False,\n",
    "              scan_start=None,\n",
    "              scan_stop=None,\n",
    "              ):\n",
    "\n",
    "    # load gwss data\n",
    "    pwindows, gwindows, _, h12, _, _ = load_h12_gwss(pop, chromosome)\n",
    "    signal = h12\n",
    "    ppos = pwindows.mean(axis=1)\n",
    "    gpos = gwindows.mean(axis=1)\n",
    "\n",
    "    # filter outliers\n",
    "    signal_filtered = hampel_filter(signal, size=filter_size, t=filter_t)\n",
    "\n",
    "    # diagnostics\n",
    "    if diagnostics:\n",
    "        fig, axs = plt.subplots(nrows=2, ncols=1, figsize=(10, 4), facecolor='w')\n",
    "        axs[0].plot(gpos, signal, marker='o', linestyle=' ', markersize=1)\n",
    "        axs[0].set_title('Original signal')\n",
    "        axs[1].plot(gpos, signal_filtered, marker='o', linestyle=' ', markersize=1)\n",
    "        axs[1].set_title('Filtered signal')\n",
    "        fig.tight_layout()\n",
    "        plt.show()\n",
    "        plt.close()\n",
    "\n",
    "    # set parameters\n",
    "    init_baseline = np.median(signal_filtered)\n",
    "    max_baseline = np.percentile(signal_filtered, max_baseline_percentile)\n",
    "    min_skew, init_skew, max_skew = -max_abs_skew, 0, max_abs_skew\n",
    "    \n",
    "    # setup output\n",
    "    records = []\n",
    "    \n",
    "    # iterate through genome\n",
    "    if not scan_start:\n",
    "        scan_start = 2\n",
    "    if not scan_stop:\n",
    "        scan_stop = gmap[chromosome][-1] - 2\n",
    "    for gcenter in np.arange(scan_start, scan_stop, scan_interval):\n",
    "    \n",
    "        for gflank in gflanks:\n",
    "\n",
    "            # locate region to fit\n",
    "            loc_region = slice(bisect_left(gpos, gcenter - gflank), \n",
    "                               bisect_right(gpos, gcenter + gflank))\n",
    "\n",
    "            # setup data to fit\n",
    "            x = gpos[loc_region]\n",
    "            y = signal_filtered[loc_region]\n",
    "\n",
    "            # fit peak model\n",
    "            peak_model = lmfit.Model(skewed_exponential_peak)\n",
    "            peak_params = lmfit.Parameters()\n",
    "            peak_params['center'] = lmfit.Parameter('center', vary=True, \n",
    "                                                    value=gcenter, \n",
    "                                                    min=gcenter - gflank, \n",
    "                                                    max=gcenter + gflank)\n",
    "            peak_params['amplitude'] = lmfit.Parameter('amplitude', vary=True, \n",
    "                                                       value=init_amplitude, \n",
    "                                                       min=min_amplitude, \n",
    "                                                       max=max_amplitude)\n",
    "            peak_params['decay'] = lmfit.Parameter('decay', \n",
    "                                                   vary=True, \n",
    "                                                   value=init_decay, \n",
    "                                                   min=min_decay, \n",
    "                                                   max=gflank/3)\n",
    "            peak_params['skew'] = lmfit.Parameter('skew', \n",
    "                                                  vary=True, \n",
    "                                                  value=init_skew, \n",
    "                                                  min=min_skew, \n",
    "                                                  max=max_skew)\n",
    "            peak_params['baseline'] = lmfit.Parameter('baseline', vary=True, \n",
    "                                                      value=init_baseline, \n",
    "                                                      min=min_baseline, \n",
    "                                                      max=max_baseline)\n",
    "            peak_params['ceiling'] = lmfit.Parameter('ceiling', vary=False, value=1)\n",
    "            peak_params['floor'] = lmfit.Parameter('floor', vary=False, value=0)\n",
    "            peak_result = peak_model.fit(y, x=x, params=peak_params)\n",
    "\n",
    "            # fit null model\n",
    "            null_model = lmfit.models.ConstantModel()\n",
    "            null_params = lmfit.Parameters()\n",
    "            null_params['c'] = lmfit.Parameter('c', vary=True, value=init_baseline, min=0, max=1)\n",
    "            null_result = null_model.fit(y, x=x, params=null_params)\n",
    "\n",
    "            # compute fit\n",
    "            peak_delta_i = int(null_result.aic - peak_result.aic)\n",
    "\n",
    "            # determine if we want to emit a result - we will do this if delta_i is above threshold\n",
    "            # and also only if the fitted peak center is within the scan interval - if it is beyond, then \n",
    "            # we will get a better fit in a different scan interval\n",
    "            fit_gcenter = peak_result.params['center'].value\n",
    "            peak_in_scan_interval = ((gcenter - scan_interval) < fit_gcenter < (gcenter + scan_interval))\n",
    "            \n",
    "            if peak_delta_i > min_aic and peak_in_scan_interval:\n",
    "                \n",
    "                fit_params = peak_result.params\n",
    "                fit_skew = fit_params['skew'].value\n",
    "                fit_decay = fit_params['decay'].value\n",
    "                decay_right = 2**(-fit_skew) * fit_decay\n",
    "                decay_left = 2**fit_skew * fit_decay\n",
    "                span1_gstart = fit_gcenter - 1*decay_left\n",
    "                span1_gstop = fit_gcenter + 1*decay_right\n",
    "                span2_gstart = fit_gcenter - 2*decay_left\n",
    "                span2_gstop = fit_gcenter + 2*decay_right\n",
    "                \n",
    "                # chromosome physical position\n",
    "                fit_pcenter = bisect_left(gmap[chromosome], fit_gcenter)\n",
    "                span1_pstart = bisect_left(gmap[chromosome], span1_gstart)\n",
    "                span1_pstop = bisect_right(gmap[chromosome], span1_gstop)\n",
    "                span2_pstart = bisect_left(gmap[chromosome], span2_gstart)\n",
    "                span2_pstop = bisect_right(gmap[chromosome], span2_gstop)\n",
    "                \n",
    "                # chromosome arm physical position\n",
    "                chromosome_arm, fit_pcenter_arm = chrom2arm(chromosome, fit_pcenter)\n",
    "                \n",
    "                # locus\n",
    "                locus = None\n",
    "                for gene in ir_genes + [tep1]:\n",
    "                    if gene.chromosome == chromosome:\n",
    "                        disjoint = (gene.chromosome_end < span1_pstart or gene.chromosome_start > span1_pstop)\n",
    "                        if not disjoint:\n",
    "                            locus = gene['Name']\n",
    "                for k, (c, p) in novel_loci.items():\n",
    "                    if c == chromosome and span1_pstart < p < span1_pstop:\n",
    "                        locus = k\n",
    "                \n",
    "                # max value, pos max value (genetic, physical)\n",
    "                loc_peak1 = slice(bisect_left(x, span1_gstart), \n",
    "                                  bisect_right(x, span2_gstop))\n",
    "                x_peak1 = x[loc_peak1]\n",
    "                y_peak1 = y[loc_peak1]\n",
    "                loc_max = np.argmax(y_peak1)\n",
    "                gpos_max = x_peak1[loc_max]\n",
    "                ppos_max = bisect_left(gmap[chromosome], gpos_max)\n",
    "                _, ppos_max_arm = chrom2arm(chromosome, ppos_max)\n",
    "                signal_max = y_peak1[loc_max]\n",
    "                \n",
    "                record = dict(\n",
    "                    pop=pop,\n",
    "                    chromosome=chromosome,\n",
    "                    gcenter=fit_gcenter,\n",
    "                    pcenter=fit_pcenter,\n",
    "                    delta_i=peak_delta_i,\n",
    "                    signal_max=signal_max,\n",
    "                    locus=locus,\n",
    "                    chromosome_arm=chromosome_arm,\n",
    "                    pcenter_arm=fit_pcenter_arm,\n",
    "                    gpos_max=gpos_max,\n",
    "                    ppos_max=ppos_max,\n",
    "                    ppos_max_arm=ppos_max_arm,\n",
    "                    span1_gstart=span1_gstart,\n",
    "                    span1_gstop=span1_gstop,\n",
    "                    span2_gstart=span2_gstart,\n",
    "                    span2_gstop=span2_gstop,\n",
    "                    span1_pstart=span1_pstart,\n",
    "                    span1_pstop=span1_pstop,\n",
    "                    span2_pstart=span2_pstart,\n",
    "                    span2_pstop=span2_pstop,\n",
    "                    amplitude=fit_params['amplitude'].value,\n",
    "                    decay=fit_decay,\n",
    "                    skew=fit_skew,\n",
    "                    decay_left=decay_left,\n",
    "                    decay_right=decay_right,\n",
    "                    baseline=fit_params['baseline'].value,\n",
    "                    aic=peak_result.aic,\n",
    "                    bic=peak_result.bic,\n",
    "                    rss=peak_result.chisqr,\n",
    "                    constant_aic=null_result.aic,\n",
    "#                     params=fit_params,\n",
    "#                     result=peak_result,\n",
    "                )\n",
    "                records.append(record)\n",
    "                \n",
    "                if diagnostics:\n",
    "                    fig, ax = plt.subplots(facecolor='w', figsize=(8, 4))\n",
    "                    peak_result.plot_fit(\n",
    "                        ax=ax, \n",
    "                        xlabel=f'Chromosome {chromosome} position (cM)', \n",
    "                        ylabel='$H12$',\n",
    "                        data_kws=dict(markersize=2), \n",
    "                        fit_kws=dict(color='k', linestyle='--')\n",
    "                    )\n",
    "                    ax.axvline(gcenter, color='w', lw=2, zorder=0)\n",
    "                    ax.axvspan(fit_gcenter - decay_left, fit_gcenter + decay_right, zorder=0, color='red', alpha=.2)\n",
    "                    ax.axvspan(fit_gcenter - 2*decay_left, fit_gcenter + 2*decay_right, zorder=0, color='red', alpha=.2)\n",
    "                    ax.axvline(fit_gcenter, color='red', lw=2, zorder=0)\n",
    "                    ax.annotate(\n",
    "                        f'$AIC={peak_result.aic:.0f}$\\n' +\n",
    "                        f'$BIC={peak_result.bic:.0f}$\\n' +\n",
    "                        f'$\\\\chi^{2}={peak_result.chisqr:.3f}$\\n' +\n",
    "                        f'$\\\\Delta_{{i}}={null_result.aic - peak_result.aic:.0f}$',\n",
    "                        xy=(0, 1), xycoords='axes fraction',\n",
    "                        xytext=(5, -5), textcoords='offset points',\n",
    "                        va='top', ha='left', fontsize=8,\n",
    "                    )\n",
    "                    ax.set_xlim(gcenter - gflank, gcenter + gflank)\n",
    "                    fig.tight_layout()\n",
    "                    plt.show()\n",
    "                    plt.close()\n",
    "                    print(peak_result.fit_report())\n",
    "                    \n",
    "    return pd.DataFrame.from_records(records)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "def dedup_peaks(df_peaks):\n",
    "    \n",
    "    keep = list(range(len(df_peaks)))\n",
    "    for i, this in df_peaks.iterrows():\n",
    "        for j, that in df_peaks.iterrows():\n",
    "            if i != j:\n",
    "                # thank you Ned Batchelder\n",
    "                # https://nedbatchelder.com/blog/201310/range_overlap_in_two_compares.html\n",
    "                disjoint = that.span1_gstart > this.span1_gstop or that.span1_gstop < this.span1_gstart\n",
    "                if not disjoint and that.delta_i > this.delta_i:\n",
    "                    keep.remove(i)\n",
    "                    break\n",
    "    return keep\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "def analyse_population(pop, min_aic=2000, diagnostics=False):\n",
    "    \n",
    "    dfs = []\n",
    "    \n",
    "    for chromosome in '23X':\n",
    "        \n",
    "        df_peaks = df_peaks = peak_scan(pop, chromosome, min_aic=min_aic, diagnostics=diagnostics)\n",
    "        keep = dedup_peaks(df_peaks)\n",
    "        df_peaks_dedup = df_peaks.iloc[keep]\n",
    "        dfs.append(df_peaks_dedup)\n",
    "\n",
    "    df = pd.concat(dfs).reset_index(drop=True)\n",
    "    df.to_csv(here() / f'data/signals/signals_{pop}.csv')\n",
    "    \n",
    "#     with open(here() / f'tables/signals_{pop}.tex', mode='w') as f:\n",
    "#         (\n",
    "#             df\n",
    "#             [['chromosome_arm', 'pcenter_arm', 'ppos_max_arm', 'locus', 'delta_i', 'signal_max']]\n",
    "#             .fillna('-')\n",
    "#             .rename(columns={\n",
    "#                 'chromosome_arm': 'Chrom',\n",
    "#                 'pcenter_arm': '$pos(peak)$',\n",
    "#                 'ppos_max_arm': '$pos(H12_{max})$',\n",
    "#                 'locus': 'Locus',\n",
    "#                 'delta_i': '$\\\\Delta_{i}$',\n",
    "#                 'signal_max': '$H12_{max}$',\n",
    "#             })\n",
    "#             .to_latex(\n",
    "#                 f,\n",
    "#                 index=False, \n",
    "#                 escape=False,\n",
    "#                 formatters=[\n",
    "#                     None, \n",
    "#                     lambda v: '{:.2f}'.format(v/1e6),\n",
    "#                     lambda v: '{:.2f}'.format(v/1e6),\n",
    "#                     '\\\\textit{{{}}}'.format,\n",
    "#                     '{:,}'.format,\n",
    "#                     '{:.2f}'.format\n",
    "#                 ]\n",
    "#             )\n",
    "#         )    \n",
    "    \n",
    "    return df\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "analyse_population('bf_gam');"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "analyse_population('gn_gam');"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "analyse_population('cm_sav_gam');"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "analyse_population('ug_gam');"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "analyse_population('gh_gam');"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "analyse_population('ga_gam');"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "analyse_population('gq_gam');"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "analyse_population('fr_gam');"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "analyse_population('bf_col');"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "analyse_population('ci_col');"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "analyse_population('gh_col');"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "analyse_population('ao_col');"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "analyse_population('gw');"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "analyse_population('gm');"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
