{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import zarr\n",
    "from pyprojroot import here\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import allel\n",
    "import yaml\n",
    "import matplotlib.pyplot as plt\n",
    "import functools\n",
    "import seaborn as sns\n",
    "sns.set_context('paper')\n",
    "sns.set_style('darkgrid')\n",
    "import dask.array as da\n",
    "import scipy.interpolate\n",
    "import scipy.stats\n",
    "import petl as etl\n",
    "import pyfasta\n",
    "from matplotlib.ticker import MultipleLocator, FormatStrFormatter\n",
    "%matplotlib inline\n",
    "%config InlineBackend.figure_format = 'retina'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "genome_path = here() / 'data/external/vectorbase/Anopheles-gambiae-PEST_CHROMOSOMES_AgamP4.fa'\n",
    "genome = pyfasta.Fasta(str(genome_path), key_fn=lambda x: x.split()[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "chromosomes = '2', '3', 'X'\n",
    "chromosome_plen = {\n",
    "    '2': len(genome['2R']) + len(genome['2L']), \n",
    "    '3': len(genome['3R']) + len(genome['3L']), \n",
    "    'X': len(genome['X'])\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "pop_defs_path = here() / 'notebooks/gwss/pop_defs.yml'\n",
    "with open(pop_defs_path, 'rt') as f:\n",
    "    pop_defs = yaml.safe_load(f)\n",
    "pops = list(pop_defs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "h12_root_path = here() / 'data/gwss/h12/h12.zarr'\n",
    "h12_root = zarr.open_consolidated(str(h12_root_path))\n",
    "\n",
    "    \n",
    "def load_h12_gwss(pop, chromosome):\n",
    "    window_size = pop_defs[pop]['h12_window_size']\n",
    "    window_step = 200\n",
    "    grp = h12_root[f'{pop}/{window_size}/{window_step}/{chromosome}']\n",
    "    return (\n",
    "        grp['windows'][:],\n",
    "        grp['gwindows'][:],\n",
    "        grp['h1'][:],\n",
    "        grp['h12'][:],\n",
    "        grp['h123'][:],\n",
    "        grp['h2_h1'][:]\n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "ihs_root_path = here() / 'data/gwss/ihs/ihs.zarr'\n",
    "ihs_root = zarr.open_consolidated(str(ihs_root_path))\n",
    "\n",
    "\n",
    "@functools.lru_cache(maxsize=None)\n",
    "def load_ihs_gwss(pop, chromosome, window_size=200, window_step=100):\n",
    "    grp = ihs_root[f'{pop}/{chromosome}']\n",
    "    pos = grp['pos'][:]\n",
    "    ihs_std = np.fabs(grp['ihs_std'][:])\n",
    "    x = allel.moving_statistic(pos, np.mean, size=window_size, step=window_step)\n",
    "    y_max = allel.moving_statistic(ihs_std, np.max, size=window_size, step=window_step)\n",
    "    y_pc95 = allel.moving_statistic(ihs_std, lambda v: np.percentile(v, 95), size=window_size, step=window_step)\n",
    "    y_pc75 = allel.moving_statistic(ihs_std, lambda v: np.percentile(v, 75), size=window_size, step=window_step)\n",
    "    y_pc50 = allel.moving_statistic(ihs_std, np.median, size=window_size, step=window_step)\n",
    "    return x, y_max, y_pc95, y_pc75, y_pc50\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "xpehh_root_path = here() / 'data/gwss/xpehh/xpehh.zarr'\n",
    "xpehh_root = zarr.open_consolidated(str(xpehh_root_path))\n",
    "\n",
    "\n",
    "@functools.lru_cache(maxsize=None)\n",
    "def load_xpehh_gwss(pop1, pop2, chromosome, window_size=500, window_step=250):\n",
    "\n",
    "    # avoid running the same scan twice\n",
    "    orig_pop1, orig_pop2 = pop1, pop2\n",
    "    pop1, pop2 = sorted([pop1, pop2])\n",
    "    \n",
    "    grp = xpehh_root[f'{pop1}_{pop2}/{chromosome}']\n",
    "    pos = grp['pos'][:]\n",
    "    xpehh = grp['xpehh'][:]\n",
    "\n",
    "    if pop1 == orig_pop2:\n",
    "        # flip back\n",
    "        xpehh = -xpehh\n",
    "        pop1, pop2 = pop2, pop1\n",
    "\n",
    "    # centre\n",
    "    xpehh = xpehh - np.median(xpehh)\n",
    "    \n",
    "    # clip at zero to focus on selection in pop1\n",
    "    xpehh1 = np.clip(xpehh, a_min=0, a_max=None)\n",
    "    \n",
    "    x = allel.moving_statistic(pos, np.mean, size=window_size, step=window_step)\n",
    "    y_max = allel.moving_statistic(xpehh1, np.max, size=window_size, step=window_step)\n",
    "    y_pc95 = allel.moving_statistic(xpehh1, lambda v: np.percentile(v, 95), size=window_size, step=window_step)\n",
    "    y_pc75 = allel.moving_statistic(xpehh1, lambda v: np.percentile(v, 75), size=window_size, step=window_step)\n",
    "    y_pc50 = allel.moving_statistic(xpehh1, np.median, size=window_size, step=window_step)\n",
    "    \n",
    "    return x, y_max, y_pc95, y_pc75, y_pc50\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "pbs_root_path = here() / 'data/gwss/pbs/pbs.zarr'\n",
    "pbs_root = zarr.open_consolidated(str(pbs_root_path))\n",
    "\n",
    "\n",
    "def load_pbs_gwss(pop1, pop2, pop3, chromosome, window_size=500, window_step=250):\n",
    "\n",
    "    grp_path = f'/{pop1}_{pop2}_{pop3}/{window_size}/{window_step}/{chromosome}'\n",
    "    grp = pbs_root[grp_path]\n",
    "    windows = grp['windows'][:]\n",
    "    gwindows = grp['gwindows'][:]\n",
    "    pbs = grp['pbs'][:]\n",
    "    pbs_scaled = grp['pbs_scaled'][:]\n",
    "    \n",
    "    return windows, gwindows, pbs, pbs_scaled\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "def load_genes():\n",
    "\n",
    "    features_path = here() / 'data/external/vectorbase/Anopheles-gambiae-PEST_BASEFEATURES_AgamP4.12.gff3'\n",
    "    df_genes = (\n",
    "        allel.gff3_to_dataframe(\n",
    "            str(features_path),\n",
    "            attributes=['ID', 'Name', 'biotype']\n",
    "        )\n",
    "        .set_index('ID')\n",
    "        .query(\"type == 'gene' and biotype == 'protein_coding'\")\n",
    "    )\n",
    "\n",
    "    # convert to chromosomal coordinates\n",
    "    df_genes['chromosome'] = df_genes['seqid'].copy()\n",
    "    df_genes['chromosome_start'] = df_genes['start'].copy()\n",
    "    df_genes['chromosome_end'] = df_genes['end'].copy()\n",
    "    loc_2R = df_genes.seqid == '2R'\n",
    "    df_genes.loc[loc_2R, 'chromosome'] = '2'\n",
    "    loc_2L = df_genes.seqid == '2L'\n",
    "    df_genes.loc[loc_2L, 'chromosome'] = '2'\n",
    "    df_genes.loc[loc_2L, 'chromosome_start'] = df_genes.loc[loc_2L, 'start'] + len(genome['2R'])\n",
    "    df_genes.loc[loc_2L, 'chromosome_end'] = df_genes.loc[loc_2L, 'end'] + len(genome['2R'])\n",
    "    loc_3R = df_genes.seqid == '3R'\n",
    "    df_genes.loc[loc_3R, 'chromosome'] = '3'\n",
    "    loc_3L = df_genes.seqid == '3L'\n",
    "    df_genes.loc[loc_3L, 'chromosome'] = '3'\n",
    "    df_genes.loc[loc_3L, 'chromosome_start'] = df_genes.loc[loc_3L, 'start'] + len(genome['3R'])\n",
    "    df_genes.loc[loc_3L, 'chromosome_end'] = df_genes.loc[loc_3L, 'end'] + len(genome['3R'])\n",
    "    df_genes['chromosome_center'] = (df_genes['chromosome_start'] + df_genes['chromosome_end']) / 2\n",
    "    \n",
    "    return df_genes\n",
    "\n",
    "\n",
    "df_genes = load_genes()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqid</th>\n",
       "      <th>source</th>\n",
       "      <th>type</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>phase</th>\n",
       "      <th>Name</th>\n",
       "      <th>biotype</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>chromosome_start</th>\n",
       "      <th>chromosome_end</th>\n",
       "      <th>chromosome_center</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AGAP004677</th>\n",
       "      <td>2L</td>\n",
       "      <td>VectorBase</td>\n",
       "      <td>gene</td>\n",
       "      <td>157348</td>\n",
       "      <td>186936</td>\n",
       "      <td>-1</td>\n",
       "      <td>-</td>\n",
       "      <td>-1</td>\n",
       "      <td>.</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>2</td>\n",
       "      <td>61702453</td>\n",
       "      <td>61732041</td>\n",
       "      <td>61717247.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AGAP004678</th>\n",
       "      <td>2L</td>\n",
       "      <td>VectorBase</td>\n",
       "      <td>gene</td>\n",
       "      <td>203779</td>\n",
       "      <td>205293</td>\n",
       "      <td>-1</td>\n",
       "      <td>+</td>\n",
       "      <td>-1</td>\n",
       "      <td>.</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>2</td>\n",
       "      <td>61748884</td>\n",
       "      <td>61750398</td>\n",
       "      <td>61749641.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AGAP004679</th>\n",
       "      <td>2L</td>\n",
       "      <td>VectorBase</td>\n",
       "      <td>gene</td>\n",
       "      <td>207894</td>\n",
       "      <td>210460</td>\n",
       "      <td>-1</td>\n",
       "      <td>+</td>\n",
       "      <td>-1</td>\n",
       "      <td>.</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>2</td>\n",
       "      <td>61752999</td>\n",
       "      <td>61755565</td>\n",
       "      <td>61754282.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AGAP004680</th>\n",
       "      <td>2L</td>\n",
       "      <td>VectorBase</td>\n",
       "      <td>gene</td>\n",
       "      <td>271285</td>\n",
       "      <td>271815</td>\n",
       "      <td>-1</td>\n",
       "      <td>+</td>\n",
       "      <td>-1</td>\n",
       "      <td>.</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>2</td>\n",
       "      <td>61816390</td>\n",
       "      <td>61816920</td>\n",
       "      <td>61816655.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AGAP004681</th>\n",
       "      <td>2L</td>\n",
       "      <td>VectorBase</td>\n",
       "      <td>gene</td>\n",
       "      <td>358329</td>\n",
       "      <td>359280</td>\n",
       "      <td>-1</td>\n",
       "      <td>-</td>\n",
       "      <td>-1</td>\n",
       "      <td>.</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>2</td>\n",
       "      <td>61903434</td>\n",
       "      <td>61904385</td>\n",
       "      <td>61903909.5</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           seqid      source  type   start     end  score strand  phase Name  \\\n",
       "ID                                                                             \n",
       "AGAP004677    2L  VectorBase  gene  157348  186936     -1      -     -1    .   \n",
       "AGAP004678    2L  VectorBase  gene  203779  205293     -1      +     -1    .   \n",
       "AGAP004679    2L  VectorBase  gene  207894  210460     -1      +     -1    .   \n",
       "AGAP004680    2L  VectorBase  gene  271285  271815     -1      +     -1    .   \n",
       "AGAP004681    2L  VectorBase  gene  358329  359280     -1      -     -1    .   \n",
       "\n",
       "                   biotype chromosome  chromosome_start  chromosome_end  \\\n",
       "ID                                                                        \n",
       "AGAP004677  protein_coding          2          61702453        61732041   \n",
       "AGAP004678  protein_coding          2          61748884        61750398   \n",
       "AGAP004679  protein_coding          2          61752999        61755565   \n",
       "AGAP004680  protein_coding          2          61816390        61816920   \n",
       "AGAP004681  protein_coding          2          61903434        61904385   \n",
       "\n",
       "            chromosome_center  \n",
       "ID                             \n",
       "AGAP004677         61717247.0  \n",
       "AGAP004678         61749641.0  \n",
       "AGAP004679         61754282.0  \n",
       "AGAP004680         61816655.0  \n",
       "AGAP004681         61903909.5  "
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_genes.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "import warnings"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "with warnings.catch_warnings():\n",
    "    warnings.simplefilter('ignore')\n",
    "    ace1 = df_genes.loc['AGAP001356']\n",
    "    ace1['Name'] = 'Ace1'\n",
    "    cyp6p3 = df_genes.loc['AGAP002865']\n",
    "    cyp6p3['Name'] = 'Cyp6p3'\n",
    "    vgsc = df_genes.loc['AGAP004707']\n",
    "    vgsc['Name'] = 'Vgsc'\n",
    "    gaba = df_genes.loc['AGAP006028']\n",
    "    gaba['Name'] = 'Gaba'\n",
    "    gste2 = df_genes.loc['AGAP009194']\n",
    "    gste2['Name'] = 'Gste2'\n",
    "    cyp9k1 = df_genes.loc['AGAP000818']\n",
    "    cyp9k1['Name'] = 'Cyp9k1'\n",
    "ir_genes = [ace1, cyp6p3, vgsc, gaba, gste2, cyp9k1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "novel_loci = {\n",
    "    'A': ('2', 24_860_000),\n",
    "    'B': ('2', 40_940_000),\n",
    "    'C': ('2', 28_549_590 + len(genome['2R'])),\n",
    "    'D': ('2', 34_050_000 + len(genome['2R'])),\n",
    "    'E': ('X', 4_360_000),\n",
    "    'F': ('X', 9_220_000),\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
